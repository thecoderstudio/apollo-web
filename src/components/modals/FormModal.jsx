import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import media from '../../util/media';
import OutlinedButton from '../buttons/OutlinedButton';
import Button from '../buttons/Button';
import Card from '../Card';
import ModalOverlay from './ModalOverlay';

const propTypes = {
  title: PropTypes.string.isRequired,
  primaryActionTitle: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired,
  onClose: PropTypes.func
};

const defaultProps = {
  onClose: () => {}
};

const Container = styled(Card)`
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  margin-left: auto;
  margin-right: auto;
  min-height: 250px;
  width: 100%;
  max-width: 400px;
  transform: translateY(-50%);

  box-sizing: border-box;
`;

const Title = styled.h2`
  text-align: center;
`;

const Form = styled.form`
  height: 100%;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 20px;
`;

const Buttons = styled.div`
  display: flex;
  justify-content: space-around;

  &> button {
    min-width: 150px;
    margin: 16px 0px;
  }

  ${
    media.phone`
      justify-content: space-between;
      flex-direction: column-reverse;
    `
  }
`;

class FormModal extends React.PureComponent {
  constructor(props) {
    super(props);
    this.close = this.close.bind(this);
  }

  close() {
    this.props.onClose();
  }

  render() {
    const {
      children,
      title,
      primaryActionTitle,
      onClose,
      ...rest
    } = this.props;

    return (
      <ModalOverlay closeModalFunction={this.close}>
        <Container>
          <Title>{title}</Title>
          <Formik validateOnChange={false} {...rest}>
            {({ values, errors, handleChange, handleSubmit }) => (
              <Form onSubmit={handleSubmit}>
                {children(values, errors, handleChange)}
                <Buttons>
                  <OutlinedButton onClick={this.close}>Cancel</OutlinedButton>
                  <Button>{primaryActionTitle}</Button>
                </Buttons>
              </Form>
            )}
          </Formik>
        </Container>
      </ModalOverlay>
    );
  }
}

FormModal.propTypes = propTypes;
FormModal.defaultProps = defaultProps;

export default FormModal;
