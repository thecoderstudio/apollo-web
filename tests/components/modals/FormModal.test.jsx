import React from 'react';
import renderer from 'react-test-renderer';
import { Formik } from 'formik';
import FormModal from '../../../src/components/modals/FormModal';

function getComponent(onClose, onSubmit) {
  return renderer.create(
    <FormModal
        title="Test"
        primaryActionTitle="Create"
        initialValues={{ test: '' }}
        onClose={onClose}
        onSubmit={onSubmit}>
        {(values, errors, handleChange) => {
          expect(values.test).toEqual('');
          return <p>Test form</p>;
        }}
    </FormModal>
  );
}

describe("form modal", () => {
  it("renders correctly", () => {
    const tree = getComponent().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("closes", () => {
    const spy = jest.fn();
    const component = getComponent(spy);
    component.root.findAllByType('button')[0].props.onClick();
    expect(spy).toHaveBeenCalled();
  });

  it("submits", () => {
    const spy = jest.fn();
    const component = getComponent(null, spy);
    component.root.findAllByType(Formik)[0].props.onSubmit();
    expect(spy).toHaveBeenCalled();
  });
});
