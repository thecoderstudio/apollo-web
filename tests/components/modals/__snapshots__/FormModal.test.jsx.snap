// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form modal renders correctly 1`] = `
.c6 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c6:hover {
  opacity: 0.90;
}

.c6:focus {
  outline: none;
}

.c6:active {
  opacity: 0.80;
}

.c6:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c7 {
  background-color: transparent;
  border: 1px solid;
}

.c1 {
  padding: 40px;
  border-radius: 10px;
}

.c0 {
  display: inline;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10;
}

.c2 {
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  margin-left: auto;
  margin-right: auto;
  min-height: 250px;
  width: 100%;
  max-width: 400px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  box-sizing: border-box;
}

.c3 {
  text-align: center;
}

.c4 {
  height: 100%;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 20px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c5> button {
  min-width: 150px;
  margin: 16px 0px;
}

@media (max-width:47.9375em) {
  .c5 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
  }
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1 c2"
  >
    <h2
      className="c3"
    >
      Test
    </h2>
    <form
      className="c4"
      onSubmit={[Function]}
    >
      <p>
        Test form
      </p>
      <div
        className="c5"
      >
        <button
          className="c6 c7"
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          className="c6"
        >
          Create
        </button>
      </div>
    </form>
  </div>
</div>
`;
